File: src/models/mutation/InversionMutation.ts
import IMutation from "./IMutation";
import Individual from "../Individual";
import _ from "lodash";

export default class InversionMutation implements IMutation {
    evaluate(individual: Individual): Individual {
        let firstIndex = ~~(Math.random() * individual.towns.length);
        let secondIndex = ~~(Math.random() * individual.towns.length);

        if (firstIndex > secondIndex)
            [firstIndex, secondIndex] = [secondIndex, firstIndex];

        const clone = _.cloneDeep(individual);

        const subList = individual.towns
            .slice(firstIndex, secondIndex)
            .reverse();
        const indicesSubSet = new Set(subList.map(town => town.index));

        clone.towns = clone.towns.filter(
            town => !indicesSubSet.has(town.index)
        );

        clone.towns.push(...subList);

        return clone;
    }
}

File: zad1/Algorithms/Mutation/InversionMutation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.VisualBasic;
using zad1.Models;

namespace zad1.Algorithms.Mutation
{
    public class InversionMutation : IMutation
    {
        public Individual Evaluate(Individual individual)
        {
            var newIndividual = new Individual(new List<Town>(individual.Towns));

            var rng = new Random();
            var firstIndex = rng.Next(0, newIndividual.Towns.Count);
            var secondIndex = rng.Next(0, newIndividual.Towns.Count);

            if (firstIndex > secondIndex)
            {
                var temp = firstIndex;
                firstIndex = secondIndex;
                secondIndex = temp;
            }

            var tempList = new List<Town>();

            for (var i = firstIndex; i <= secondIndex; i++) tempList.Add(newIndividual.Towns[i]);

            tempList.Reverse();

            for (var i = firstIndex; i <= secondIndex; i++) newIndividual.Towns[i] = tempList[i - firstIndex];

            return newIndividual;
        }
    }
}